Ensure UDP and DTLS transports can handle message sizes up to 64KB

bwalden opened the pull request:
Fixes #238.

sgodin commented on resip/stack/ssl/DtlsTransport.cxx:
```diff
@@ -167,9 +167,8 @@ DtlsTransport::_read( FdSet& fdset )
-   unsigned int bufferLen = UdpTransport::MaxBufferSize + 5 ;
-   char* buffer = new char[ bufferLen ] ;
-   unsigned char *pt = new unsigned char[ bufferLen ] ;
+   std::unique_ptr<char[]> buffer(MsgHeaderScanner::allocateBuffer(MaxMessageSize));
```
So we are allocating 2 64K buffers for every _read operation?  Might it make sense to make these class members and resuse them (like in UdpTransport)?  I see a copy operation below, so this is using the copy solution.

bwalden commented on resip/stack/UdpTransport.cxx:
```diff
@@ -530,24 +490,26 @@ UdpTransport::processRxParse(char *buffer, int len, Tuple& sender)
 #ifdef USE_SIGCOMP
-      char* newBuffer = MsgHeaderScanner::allocateBuffer(MaxBufferSize);
-      size_t uncompressedLength = mSigcompStack->uncompressMessage(buffer, len, newBuffer, MaxBufferSize, sc);
+      char* newBuffer = MsgHeaderScanner::allocateBuffer(MaxMessageSize);
```
Updated to use an additional 64K buffer per UdpTransport if USE_SIGCOMP is defined. This probably isn't ideal, but not being familiar with OpenSigComp, I couldn't find a way to trivially determine the compression factor.

bwalden commented:
## Before #239:
```
Performing 10000 runs with win=100 ipv4 proto=udp numports=1 thread=event bindIf=127.0.0.1 listen=1 tf=0.
WARNING | 20220506-090736.587 | resip_test.exe | RESIP | 51444 | Security.cxx:3126 | unable to load DH parameters (required for PFS): TlsDHParamsFilename not specified
WARNING | 20220506-090736.594 | resip_test.exe | RESIP | 51444 | Security.cxx:3126 | unable to load DH parameters (required for PFS): TlsDHParamsFilename not specified
WARNING | 20220506-090736.594 | resip_test.exe | RESIP | 51444 | Security.cxx:346 | No root certificates found using legacy prefixes, treating mPath as a normal directory of root certs
WARNING | 20220506-090736.653 | resip_test.exe | RESIP | 51444 | Security.cxx:3126 | unable to load DH parameters (required for PFS): TlsDHParamsFilename not specified
WARNING | 20220506-090736.655 | resip_test.exe | RESIP | 51444 | Security.cxx:3126 | unable to load DH parameters (required for PFS): TlsDHParamsFilename not specified
WARNING | 20220506-090736.655 | resip_test.exe | RESIP | 51444 | Security.cxx:346 | No root certificates found using legacy prefixes, treating mPath as a normal directory of root certs
WARNING | 20220506-090836.583 | resip_test.exe | RESIP:STATS | 24024 | StatisticsMessage.cxx:98 | RESIP:TRANSACTION
TU summary: 0 TRANSPORT 2 TRANSACTION 1 CLIENTTX 0 SERVERTX 1723 TIMERS 1911
Transaction summary: reqi 6196 reqo 0 rspi 0 rspo 3196
Details: INVi 0/S0/F0 INVo 0/S0/F0 ACKi 0 ACKo 0 BYEi 0/S0/F0 BYEo 0/S0/F0 CANi 0/S0/F0 CANo 0/S0/F0 MSGi 0/S0/F0 MSGo 0/S0/F0 OPTi 0/S0/F0 OPTo 0/S0/F0 REGi 6196/S3196/F0 REGo 0/S0/F0 PUBi 0/S0/F0 PUBo 0/S0/F0 SUBi 0/S0/F0 SUBo 0/S0/F0 NOTi 0/S0/F0 NOTo 0/S0/F0 REFi 0/S0/F0 REFo 0/S0/F0 INFi 0/S0/F0 INFo 0/S0/F0 PRAi 0/S0/F0 PRAo 0/S0/F0 SERi 0/S0/F0 SERo 0/S0/F0 UPDi 0/S0/F0 UPDo 0/S0/F0
Retransmissions: INVx 0 finx 0 nonx 0 BYEx 0 CANx 0 MSGx 0 OPTx 0 REGx 0 PUBx 0 SUBx 0 NOTx 0 REFx 0 INFx 0 PRAx 0 SERx 0 UPDx 0
WARNING | 20220506-090836.663 | resip_test.exe | RESIP:STATS | 32028 | StatisticsMessage.cxx:98 | RESIP:TRANSACTION
TU summary: 0 TRANSPORT 0 TRANSACTION 1 CLIENTTX 369 SERVERTX 0 TIMERS 2187
Transaction summary: reqi 0 reqo 3296 rspi 6201 rspo 0
Details: INVi 0/S0/F0 INVo 0/S0/F0 ACKi 0 ACKo 0 BYEi 0/S0/F0 BYEo 0/S0/F0 CANi 0/S0/F0 CANo 0/S0/F0 MSGi 0/S0/F0 MSGo 0/S0/F0 OPTi 0/S0/F0 OPTo 0/S0/F0 REGi 0/S0/F0 REGo 4294964129/S6201/F0 PUBi 0/S0/F0 PUBo 0/S0/F0 SUBi 0/S0/F0 SUBo 0/S0/F0 NOTi 0/S0/F0 NOTo 0/S0/F0 REFi 0/S0/F0 REFo 0/S0/F0 INFi 0/S0/F0 INFo 0/S0/F0 PRAi 0/S0/F0 PRAo 0/S0/F0 SERi 0/S0/F0 SERo 0/S0/F0 UPDi 0/S0/F0 UPDo 0/S0/F0
Retransmissions: INVx 0 finx 0 nonx 0 BYEx 0 CANx 0 MSGx 0 OPTx 0 REGx 6463 PUBx 0 SUBx 0 NOTx 0 REFx 0 INFx 0 PRAx 0 SERx 0 UPDx 0
WARNING | 20220506-090936.596 | resip_test.exe | RESIP:STATS | 24024 | StatisticsMessage.cxx:98 | RESIP:TRANSACTION
TU summary: 0 TRANSPORT 4 TRANSACTION 1 CLIENTTX 0 SERVERTX 1943 TIMERS 2159
Transaction summary: reqi 13067 reqo 0 rspi 0 rspo 6768
Details: INVi 0/S0/F0 INVo 0/S0/F0 ACKi 0 ACKo 0 BYEi 0/S0/F0 BYEo 0/S0/F0 CANi 0/S0/F0 CANo 0/S0/F0 MSGi 0/S0/F0 MSGo 0/S0/F0 OPTi 0/S0/F0 OPTo 0/S0/F0 REGi 13067/S6768/F0 REGo 0/S0/F0 PUBi 0/S0/F0 PUBo 0/S0/F0 SUBi 0/S0/F0 SUBo 0/S0/F0 NOTi 0/S0/F0 NOTo 0/S0/F0 REFi 0/S0/F0 REFo 0/S0/F0 INFi 0/S0/F0 INFo 0/S0/F0 PRAi 0/S0/F0 PRAo 0/S0/F0 SERi 0/S0/F0 SERo 0/S0/F0 UPDi 0/S0/F0 UPDo 0/S0/F0
Retransmissions: INVx 0 finx 0 nonx 0 BYEx 0 CANx 0 MSGx 0 OPTx 0 REGx 0 PUBx 0 SUBx 0 NOTx 0 REFx 0 INFx 0 PRAx 0 SERx 0 UPDx 0
WARNING | 20220506-090936.654 | resip_test.exe | RESIP:STATS | 32028 | StatisticsMessage.cxx:98 | RESIP:TRANSACTION
TU summary: 0 TRANSPORT 0 TRANSACTION 0 CLIENTTX 408 SERVERTX 0 TIMERS 2465
Transaction summary: reqi 0 reqo 6866 rspi 13065 rspo 0
Details: INVi 0/S0/F0 INVo 0/S0/F0 ACKi 0 ACKo 0 BYEi 0/S0/F0 BYEo 0/S0/F0 CANi 0/S0/F0 CANo 0/S0/F0 MSGi 0/S0/F0 MSGo 0/S0/F0 OPTi 0/S0/F0 OPTo 0/S0/F0 REGi 0/S0/F0 REGo 4294961041/S13065/F0 PUBi 0/S0/F0 PUBo 0/S0/F0 SUBi 0/S0/F0 SUBo 0/S0/F0 NOTi 0/S0/F0 NOTo 0/S0/F0 REFi 0/S0/F0 REFo 0/S0/F0 INFi 0/S0/F0 INFo 0/S0/F0 PRAi 0/S0/F0 PRAo 0/S0/F0 SERi 0/S0/F0 SERo 0/S0/F0 UPDi 0/S0/F0 UPDo 0/S0/F0
Retransmissions: INVx 0 finx 0 nonx 0 BYEx 0 CANx 0 MSGx 0 OPTx 0 REGx 13121 PUBx 0 SUBx 0 NOTx 0 REFx 0 INFx 0 PRAx 0 SERx 0 UPDx 0
10000 registrations performed in 175019 ms, a rate of 57.1367 transactions per second.
WARNING | 20220506-091031.757 | resip_test.exe | RESIP:TRANSACTION | 51444 | TransactionController.cxx:72 | On shutdown, there are Client TransactionStates remaining!
WARNING | 20220506-091031.789 | resip_test.exe | RESIP:TRANSACTION | 51444 | TransactionController.cxx:77 | On shutdown, there are Server TransactionStates remaining!
```
![image](https://user-images.githubusercontent.com/5887185/167093855-1bd37052-e009-45a7-92ec-44ae8271aaff.png)


## After #239:
```
Performing 10000 runs with win=100 ipv4 proto=udp numports=1 thread=event bindIf=127.0.0.1 listen=1 tf=0.
WARNING | 20220506-081737.302 | resip_test.exe | RESIP | 47848 | Security.cxx:3126 | unable to load DH parameters (required for PFS): TlsDHParamsFilename not specified
WARNING | 20220506-081737.304 | resip_test.exe | RESIP | 47848 | Security.cxx:3126 | unable to load DH parameters (required for PFS): TlsDHParamsFilename not specified
WARNING | 20220506-081737.305 | resip_test.exe | RESIP | 47848 | Security.cxx:346 | No root certificates found using legacy prefixes, treating mPath as a normal directory of root certs
WARNING | 20220506-081737.330 | resip_test.exe | RESIP | 47848 | Security.cxx:3126 | unable to load DH parameters (required for PFS): TlsDHParamsFilename not specified
WARNING | 20220506-081737.331 | resip_test.exe | RESIP | 47848 | Security.cxx:3126 | unable to load DH parameters (required for PFS): TlsDHParamsFilename not specified
WARNING | 20220506-081737.331 | resip_test.exe | RESIP | 47848 | Security.cxx:346 | No root certificates found using legacy prefixes, treating mPath as a normal directory of root certs
WARNING | 20220506-081837.299 | resip_test.exe | RESIP:STATS | 13836 | StatisticsMessage.cxx:98 | RESIP:TRANSACTION
TU summary: 0 TRANSPORT 9 TRANSACTION 2 CLIENTTX 0 SERVERTX 4922 TIMERS 5478
Transaction summary: reqi 16669 reqo 0 rspi 0 rspo 9122
Details: INVi 0/S0/F0 INVo 0/S0/F0 ACKi 0 ACKo 0 BYEi 0/S0/F0 BYEo 0/S0/F0 CANi 0/S0/F0 CANo 0/S0/F0 MSGi 0/S0/F0 MSGo 0/S0/F0 OPTi 0/S0/F0 OPTo 0/S0/F0 REGi 16669/S9122/F0 REGo 0/S0/F0 PUBi 0/S0/F0 PUBo 0/S0/F0 SUBi 0/S0/F0 SUBo 0/S0/F0 NOTi 0/S0/F0 NOTo 0/S0/F0 REFi 0/S0/F0 REFo 0/S0/F0 INFi 0/S0/F0 INFo 0/S0/F0 PRAi 0/S0/F0 PRAo 0/S0/F0 SERi 0/S0/F0 SERo 0/S0/F0 UPDi 0/S0/F0 UPDo 0/S0/F0
Retransmissions: INVx 0 finx 0 nonx 0 BYEx 0 CANx 0 MSGx 0 OPTx 0 REGx 0 PUBx 0 SUBx 0 NOTx 0 REFx 0 INFx 0 PRAx 0 SERx 0 UPDx 0
WARNING | 20220506-081837.330 | resip_test.exe | RESIP:STATS | 52568 | StatisticsMessage.cxx:98 | RESIP:TRANSACTION
TU summary: 0 TRANSPORT 0 TRANSACTION 1 CLIENTTX 891 SERVERTX 0 TIMERS 5929
Transaction summary: reqi 0 reqo 9218 rspi 16663 rspo 0
Details: INVi 0/S0/F0 INVo 0/S0/F0 ACKi 0 ACKo 0 BYEi 0/S0/F0 BYEo 0/S0/F0 CANi 0/S0/F0 CANo 0/S0/F0 MSGi 0/S0/F0 MSGo 0/S0/F0 OPTi 0/S0/F0 OPTo 0/S0/F0 REGi 0/S0/F0 REGo 67/S16663/F0 PUBi 0/S0/F0 PUBo 0/S0/F0 SUBi 0/S0/F0 SUBo 0/S0/F0 NOTi 0/S0/F0 NOTo 0/S0/F0 REFi 0/S0/F0 REFo 0/S0/F0 INFi 0/S0/F0 INFo 0/S0/F0 PRAi 0/S0/F0 PRAo 0/S0/F0 SERi 0/S0/F0 SERo 0/S0/F0 UPDi 0/S0/F0 UPDo 0/S0/F0
Retransmissions: INVx 0 finx 0 nonx 0 BYEx 0 CANx 0 MSGx 0 OPTx 0 REGx 9151 PUBx 0 SUBx 0 NOTx 0 REFx 0 INFx 0 PRAx 0 SERx 0 UPDx 0
10000 registrations performed in 66790 ms, a rate of 149.723 transactions per second.
WARNING | 20220506-081844.162 | resip_test.exe | RESIP:TRANSACTION | 47848 | TransactionController.cxx:72 | On shutdown, there are Client TransactionStates remaining!
WARNING | 20220506-081844.166 | resip_test.exe | RESIP:TRANSACTION | 47848 | TransactionController.cxx:77 | On shutdown, there are Server TransactionStates remaining!
WARNING | 20220506-081844.171 | resip_test.exe | RESIP:TRANSPORT | 47848 | InternalTransport.cxx:60 | TX Fifo non-empty in ~InternalTransport! Has 5 messages.
```
![image](https://user-images.githubusercontent.com/5887185/167088072-9f8be91b-498a-4a5a-9084-972382ee8014.png)

sgodin commented:
Hey Ben, sorry I've been swamped lately.  Looks like transactions per second actually improved, nice!  I'll try to spend some time with this today.

sgodin commented:
Okay - looks good to me.  Two things before I think we are good to merge:
1.  We should remove the definition of RESIP_TRANSPORT_FLAG_KEEP_BUFFER from Transport.hxx (and from the comments above it).
2. Due to the extensive changes in DTLS (which is great removal of code duplication) - I'd like to know that the DTLS transport has been tested.  Have you tested it?

sgodin commented on resip/stack/UdpTransport.cxx:
```diff
@@ -321,72 +316,43 @@ UdpTransport::processTxOne(SendData *data)
 void
```
Note:  There are references to the removed KEEP_BUFFER flag in the above comment, we need to clean up.

bwalden commented:
Hi @sgodin that's for the review comments. I found some time today to action them.

1. I've removed the RESIP_TRANSPORT_FLAG_KEEP_BUFFER flag as requested. It doesn't look like anything else references it.
2. I tested the DTLS changes within my project by making a few calls between two endpoints using DTLS. All appeared to be OK. I realised that test stack didn't have the ability to exercise TLS or DTLS, so I've made some changes to add those transport options. All appears to be OK using those transports with test stack too. Although I did run into some issues in Security.cxx before I'd set up the certs correctly. There were instances where catching BaseSecurity::Exception would cause other exceptions to drop through to the catch-all, losing the error message. I've changed these to catch BaseException where I thought it was appropriate.

sgodin commented:
That all sounds amazing - thanks Ben.

Unfortunately (for this PR), I am on vacation now and all next week, but I
hope to get back to reviewing this after I return.

Thanks for all your great contributions.

Best Regards,
Scott

On Fri, Jun 3, 2022 at 12:34 PM Ben Walden ***@***.***> wrote:

> Hi @sgodin <https://github.com/sgodin> that's for the review comments. I
> found some time today to action them.
>
>    1. I've removed the RESIP_TRANSPORT_FLAG_KEEP_BUFFER flag as
>    requested. It doesn't look like anything else references it.
>    2. I tested the DTLS changes within my project by making a few calls
>    between two endpoints using DTLS. All appeared to be OK. I realised that
>    test stack didn't have the ability to exercise TLS or DTLS, so I've made
>    some changes to add those transport options. All appears to be OK using
>    those transports with test stack too. Although I did run into some issues
>    in Security.cxx before I'd set up the certs correctly. There were instances
>    where catching BaseSecurity::Exception would cause other exceptions to drop
>    through to the catch-all, losing the error message. I've changed these to
>    catch BaseException where I thought it was appropriate.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/resiprocate/resiprocate/pull/239#issuecomment-1146159548>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVU2UNPYNNRFSN2TWZZVITVNIXZPANCNFSM5VAJGP4Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
